router/internal/httpclient/traceclient.go

```diff
@@ -0,0 +1,211 @@
+				t.logger.Error("failed to marshal Got1xxResponse", zap.Error(err))
+				t.traceAndLog("marshal error", operation, context, "")
+				return nil
+			}
+
+			t.traceAndLog(string(baseBytes), operation, context, "")
+			return nil
+		},
+		DNSStart: func(dnsStartInfo httptrace.DNSStartInfo) {
+			t.traceAndLog(dnsStartInfo.Host, "DNSStart", context, otel.DnsQuestionName)
+		},
+		DNSDone: func(dnsDoneInfo httptrace.DNSDoneInfo) {
+			operationString := "DNSDone"
+			baseBytes, err := json.Marshal(dnsDoneInfo)
+			if err != nil {
+				t.logger.Error("failed to marshal DNSDone", zap.Error(err))
+				t.traceAndLog("marshal error", operationString, context, "")
+			}
+			t.traceAndLog(string(baseBytes), operationString, context, "")
+		},
+		ConnectStart: func(network, addr string) {
+			baseString := fmt.Sprintf(`{"network":%q,"addr":%q}`, network, addr)
+			t.traceAndLog(baseString, "ConnectStart", context, "")
+		},
+		ConnectDone: func(network, addr string, err error) {
+			baseString := fmt.Sprintf(`{"network":%q,"addr":%q,"err":%s}`, network, addr, t.getErrString(err))
+			t.traceAndLog(baseString, "ConnectDone", context, "")
+		},
+		TLSHandshakeStart: func() {
+			t.traceAndLog(true, "TLSHandshakeStart", context, "")
+		},
+		TLSHandshakeDone: func(connectionState tls.ConnectionState, err error) {
+			t.trace(t.getCipherSuiteName(connectionState.CipherSuite), "", context, otel.TlsCipher)
+			value := fmt.Sprintf(`{ "tls": { "handshakeComplete":%t }, "error":%s}`, connectionState.HandshakeComplete, t.getErrString(err))
+			t.traceAndLog(value, "TLSHandshakeDone", context, "")
+		},
+		// We are skipping since this runs for individual fields multiple times
+		// and we can rely on WroteHeaders instead
+		WroteHeaderField: nil,
+		WroteHeaders: func() {
+			t.traceAndLog(true, "WroteHeaders", context, "")
+		},
+		Wait100Continue: func() {
+			t.traceAndLog(true, "Wait100Continue", context, "")
+		},
+		WroteRequest: func(wroteRequestInfo httptrace.WroteRequestInfo) {
+			baseString := fmt.Sprintf(`{ "error": %s }`, t.getErrString(wroteRequestInfo.Err))
+			t.traceAndLog(baseString, "WroteRequest", context, "")
+		},
+	}
+	return trace
+}
+
+func (t *TraceInjectingRoundTripper) getErrString(err error) string {
+	errString := "null"
+	if err != nil {
+		errString = fmt.Sprintf("%q", err)
+	}
+	return errString
+}
+
+// Helper to get the cipher suite to follow the standard trace attribute
+func (t *TraceInjectingRoundTripper) getCipherSuiteName(id uint16) string {
+	for _, cs := range tls.CipherSuites() {
+		if cs.ID == id {
+			return cs.Name
+		}
+	}
+	for _, cs := range tls.InsecureCipherSuites() {
+		if cs.ID == id {
+			return cs.Name
+		}
+	}
+	return fmt.Sprintf("%d", id)
+}
+
+// We only use this for Got1xxResponse response at the moment
+// it can also be used for WroteHeaderField if we enable it later
+func (t *TraceInjectingRoundTripper) addAndGetAttrVal(value map[string]any, operationString string) []any {
+	t.mu.Lock()
```

Comment:
We create a client trace transport per request, but even within the same request some operations can get called multiple times. From the code

> Any particular hook may be nil. Functions may be
> called concurrently from different goroutines and some may be called
> after the request has completed or failed.

Notably `Got1xxResponse` and `WroteHeaderField`, of which the latter we are skipping since we have `WroteHeaders`.